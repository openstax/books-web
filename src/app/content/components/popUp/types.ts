import { LinkedArchiveTree, LinkedArchiveTreeNode, LinkedArchiveTreeSection } from '../../types';

export type LocationFilters = Map<
  string,
  { section: LinkedArchiveTreeNode, children?: LinkedArchiveTreeSection[] }
>;

export type LocationFiltersWithChildren = Map<
  string,
  { section: LinkedArchiveTree, children: LinkedArchiveTreeSection[] }
>;

export interface FiltersChange<T> {
  remove: T[];
  new: T[];
}
