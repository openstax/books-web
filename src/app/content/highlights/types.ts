import { Highlight, HighlightsSummary } from '@openstax/highlighter/dist/api';

export type HighlightData = Highlight;

export type CountsPerSource = Exclude<HighlightsSummary['countsPerSource'], undefined>;

export interface State {
  myHighlightsOpen: boolean;
  enabled: boolean;
  focused?: string;
  highlights: null | HighlightData[];
  summary: {
    filters: {
      colors: string[];
      chapters: string[];
    },
    pagination: {
      // even though we're manually splitting our requests out into smaller batches of sources,
      // if a source has more than [numPage] highlights the api will still paginate. in order to
      // access the subsequent pages, we need to store the original sources here, and continue to
      // use them until we run out of pages, then grab the next batch of sources. on the LAST
      // page the response size may be smaller than expected, in this case the next batch of
      // sources should be immediately queried to fill the remaining [numPage]. use
      // paginationUtils.getNextPageSources to get the next batch of sources
      sources: string[];
      page: number;
    } | null,
    loading: boolean;
    // totalCounts should reflect the UNFILTERED state of user's data, used in representing
    // available filter options. this should be updated on create/delete and after auth change
    totalCounts: {[key: string]: number};
    // filteredTotalCounts should reflect the FILTERED state of user's data, used in calculating
    // pagination requests. this should be updated after filter change before requesting
    // filtered highlights (because you need this to query them)
    filteredTotalCounts: {[key: string]: number};
    highlights: {
      [chapterId: string]: {[pageId: string]: HighlightData[]}
    };
  };
}
