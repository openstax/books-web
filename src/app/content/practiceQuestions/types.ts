import { LinkedArchiveTree, LinkedArchiveTreeSection } from '../types';

export interface State {
  isEnabled: boolean;
  summary: PracticeQuestionsSummary | null;
  open: boolean;
  selectedSection: LinkedArchiveTreeSection | null;
  currentQuestionIndex: number | null;
  questions: PracticeQuestion[];
}

export interface PracticeQuestionsSummary {
  countsPerSource: {
    [key: string]: number;
  };
}

interface PracticeAnswer {
  id: number;
  content_html: string;
  correctness: '0.0' | '1.0';
  feedback_html: string;
}

export interface PracticeQuestion {
  group_uuid: string;
  uid: string;
  stem_html: string;
  answers: PracticeAnswer[];
}

export type PracticeQuestions = PracticeQuestion[];

export type PracticeQuestionsLocationFilters = Map<
  string,
  { section: LinkedArchiveTree, children: LinkedArchiveTreeSection[] }
>;
