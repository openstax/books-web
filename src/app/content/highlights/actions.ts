import { createStandardAction } from 'typesafe-actions';

export const enableHighlighting = createStandardAction('FeatureFlag/enableHighlighting')<void>();
