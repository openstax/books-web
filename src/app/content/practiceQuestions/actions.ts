import { createStandardAction } from 'typesafe-actions';
import { PracticeQuestionsSummary } from './types';

export const receivePracticeQuestionsSummary = createStandardAction(
  'Content/PracticeQuestions/Summary/receive'
)<PracticeQuestionsSummary>();
