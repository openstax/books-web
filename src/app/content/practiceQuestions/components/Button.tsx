import styled from 'styled-components/macro';
import { PlainButton } from '../../../components/Button';
import theme from '../../../theme';

interface ButtonProps {
  withoutBackground: boolean;
}

export default styled(PlainButton)`
  backgrouond-color: ${theme.color.primary.orange.base};
  color: ${theme.color.text.white};
  font-size: 1.6rem;
  font-weight: bold;
  height: 5rem;
  padding: 0 3rem;
  ${(props: ButtonProps) => {
    if (props.withoutBackground) {
      return `
        backgrouond-color: transparent;
        color: #027eb5;
      `;
    }
  }}
`;
