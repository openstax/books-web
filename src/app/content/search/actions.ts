import { createStandardAction } from 'typesafe-actions';
import { State } from './types';

export const receiveSearchResults = createStandardAction('Content/Search/receiveResults')<
  Exclude<State['results'], null>
>();
export const requestSearch = createStandardAction('Content/Search/request')<string>();
export const clearSearch = createStandardAction('Content/Search/clear')();
