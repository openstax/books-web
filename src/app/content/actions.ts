import { createStandardAction } from 'typesafe-actions';
import * as highlightingActions from './highlights/actions';
import * as searchActions from './search/actions';
import { ArchivePage, Book, Params, State } from './types';

export const search = searchActions;
export const highlighting = highlightingActions;

export const openToc = createStandardAction('Content/openToc')<void>();
export const closeToc = createStandardAction('Content/closeToc')<void>();
export const resetToc = createStandardAction('Content/resetToc')<void>();

export const requestBook = createStandardAction('Content/requestBook')<Params['book']>();
export const receiveBook = createStandardAction('Content/receiveBook')<Book>();

export const requestPage = createStandardAction('Content/requestPage')<Params['page']>();
export const receivePage = createStandardAction('Content/receivePage')<
  ArchivePage & {references: State['references']}
>();

export const closeCallToActionPopup = createStandardAction('Content/closeCallToActionPopup')();
export const openCallToActionPopup = createStandardAction('Content/openCallToActionPopup')();
