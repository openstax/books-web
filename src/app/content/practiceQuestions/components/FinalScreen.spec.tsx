import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import createTestServices from '../../../../test/createTestServices';
import createTestStore from '../../../../test/createTestStore';
import Button from '../../../components/Button';
import * as Services from '../../../context/Services';
import MessageProvider from '../../../MessageProvider';
import { Store } from '../../../types';
import { LinkedArchiveTreeSection } from '../../types';
import { setSelectedSection } from '../actions';
import FinalScreen from './FinalScreen';

describe('FinalScreen for practice questions', () => {
  let store: Store;
  let dispatch: jest.SpyInstance;
  let services: ReturnType<typeof createTestServices>;

  beforeEach(() => {
    store = createTestStore();
    dispatch = jest.spyOn(store, 'dispatch');
    services = createTestServices();
  });

  it('renders properly and dispatches action on click', () => {
    const mockSection = { title: 'some title' } as LinkedArchiveTreeSection;

    const component = renderer.create(<Provider store={store}>
      <Services.Provider value={services}>
        <MessageProvider>
          <FinalScreen nextSection={mockSection} />
        </MessageProvider>
      </Services.Provider>
    </Provider>);

    const button = component.root.findByType(Button);

    renderer.act(() => {
      button.props.onClick();
    });

    expect(dispatch).toHaveBeenCalledWith(setSelectedSection(mockSection));
    expect(component.toJSON()).toMatchSnapshot();
  });
});
