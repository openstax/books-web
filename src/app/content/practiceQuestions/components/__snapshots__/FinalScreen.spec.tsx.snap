// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FinalScreen for practice questions renders properly and dispatches action on click 1`] = `
<div
  className="FinalScreen__StyledFinalScreen-sc-131hu2n-0 cKZyXt"
>
  <span
    className="FinalScreen__StyledText-sc-131hu2n-1 MXQtJ"
  >
    You've reached the end of this set
  </span>
  <div
    className="NextSectionMessage__StyledNextSectionMessage-qtwzgn-0 gTWiOh"
  >
    <div
      className="NextSectionMessage__StyledMessage-qtwzgn-1 TEURJ"
    >
      <span
        className="NextSectionMessage__StyledMessageText-qtwzgn-2 iCtbqd"
      >
        Select a section from the dropdown above or continue to 
      </span>
      <span
        className="NextSectionMessage__StyledSectionTitle-qtwzgn-3 cmlWLO"
        dangerouslySetInnerHTML={
          Object {
            "__html": "some title",
          }
        }
      />
    </div>
    <button
      className="Button-ayg7nk-0 crGJLq"
      onClick={[Function]}
    >
      Continue
    </button>
  </div>
</div>
`;
